libdnf/advisory/advisory_query.cpp

```diff
@@ -0,0 +1,437 @@
+        }
+
+        switch (tmp_cmp_type) {
+            case libdnf::sack::QueryCmp::EQ: {
+                std::string prefixed_name = std::string(libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX) + pattern;
+                Id name_id = pool_str2id(pool, prefixed_name.c_str(), 0);
+                for (Id candidate_id : p_impl->data_map) {
+                    Id candidate_name = pool_lookup_id(pool, candidate_id, SOLVABLE_NAME);
+                    if (candidate_name == name_id) {
+                        filter_result.add_unsafe(candidate_id);
+                    }
+                }
+            } break;
+            case libdnf::sack::QueryCmp::GLOB: {
+                for (Id candidate_id : p_impl->data_map) {
+                    const char * candidate_name = pool_lookup_str(pool, candidate_id, SOLVABLE_NAME);
+                    if (fnmatch(c_pattern, candidate_name + libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH, 0) ==
+                        0) {
+                        filter_result.add_unsafe(candidate_id);
+                    }
+                }
+
+            } break;
+            case libdnf::sack::QueryCmp::IGLOB: {
+                for (Id candidate_id : p_impl->data_map) {
+                    const char * candidate_name = pool_lookup_str(pool, candidate_id, SOLVABLE_NAME);
+                    if (fnmatch(
+                            c_pattern,
+                            candidate_name + libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH,
+                            FNM_CASEFOLD) == 0) {
+                        filter_result.add_unsafe(candidate_id);
+                    }
+                }
+
+            } break;
+            default:
+                throw NotSupportedCmpType("Unsupported CmpType");
+        }
+    }
+
+    // Apply filter results to query
+    if (cmp_not) {
+        p_impl->data_map -= filter_result;
+    } else {
+        p_impl->data_map &= filter_result;
+    }
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_type(const Advisory::Type type, sack::QueryCmp cmp_type) {
+    const std::vector<Advisory::Type> types = {type};
+    ifilter_type(types, cmp_type);
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_type(const std::vector<Advisory::Type> & types, sack::QueryCmp cmp_type) {
+    auto & solv_sack = advisory_sack->base->get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+    libdnf::rpm::solv::SolvMap filter_result(solv_sack.p_impl->get_nsolvables());
+
+    bool cmp_not = (cmp_type & libdnf::sack::QueryCmp::NOT) == libdnf::sack::QueryCmp::NOT;
+    if (cmp_not) {
+        // Removal of NOT CmpType makes following comparissons easier and more effective
+        cmp_type = cmp_type - libdnf::sack::QueryCmp::NOT;
+    }
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ: {
+            for (const Advisory::Type & type : types) {
```

Comment:
(For discussion) - Do we have correctly defined `Advisory::Type`? Can we improve an iteration?

Reply:
For now I have kept both of the enums defined as a bit in int but I have unified the public API so that it always takes a vector of enum values and I have improved the iteration for `ifilter_type`.

I have also added TODOs to record we might want to convert the enums from bits in int to regular ints for the public API.